File: web/src/logic/player.ts
import { DamageType } from './types';

export class Player {
  name: string;
  maxHp: number = 100;
  hp: number;
  attackMin: number;
  attackMax: number;
  magicMin: number;
  magicMax: number;
  strength: number;
  intelligence: number;
  dexterity: number;
  potions = 3;

  constructor(init: Partial<Player> & { name: string }) {
    Object.assign(this, init);
    this.hp = this.maxHp;
  }

  get alive() {
    return this.hp > 0;
  }

  physicalAttack() {
    const base = randomInt(this.attackMin, this.attackMax);
    return base + Math.floor(this.strength / 4);
  }

  magicAttack() {
    const base = randomInt(this.magicMin, this.magicMax);
    return base + Math.floor(this.intelligence / 4);
  }

  heal() {
    if (this.potions <= 0) throw new Error('No potions');
    this.potions -= 1;
    const heal = Math.min(20, this.maxHp - this.hp);
    this.hp += heal;
    return heal;
  }
}

function randomInt(min: number, max: number) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
} 

File: ai_rpg/player.py
from __future__ import annotations

import random
from dataclasses import dataclass, field


@dataclass
class Player:
    """Represents the player character."""

    name: str
    max_hp: int = 100
    attack_min: int = 10
    attack_max: int = 20
    potion_heal_amount: int = 20
    potions: int = 3
    rng: random.Random = field(default_factory=random.Random, repr=False)

    hp: int = field(init=False)

    def __post_init__(self) -> None:
        self.hp = self.max_hp

    @property
    def alive(self) -> bool:
        return self.hp > 0

    def attack(self) -> int:
        """Calculate player's attack damage."""
        damage = self.rng.randint(self.attack_min, self.attack_max)
        return damage

    def receive_damage(self, amount: int) -> None:
        self.hp = max(self.hp - amount, 0)

    def heal(self) -> int:
        if self.potions <= 0:
            raise RuntimeError("No potions left!")
        self.potions -= 1
        healed = min(self.potion_heal_amount, self.max_hp - self.hp)
        self.hp += healed
        return healed

    def full_status(self) -> str:
        return (
            f"{self.name}: HP {self.hp}/{self.max_hp} | "
            f"Potions: {self.potions}"
        ) 